code/modules/reagents/chemistry/holder.dm

```diff
@@ -677,138 +727,320 @@
-				multiplier = max(multiplier, 1) //this shouldn't happen ...
-				SSblackbox.record_feedback("tally", "chemical_reaction", cached_results[P]*multiplier, P)
-				add_reagent(P, cached_results[P]*multiplier, null, chem_temp)
+	if(.)
+		SEND_SIGNAL(src, COMSIG_REAGENTS_REACTED, .)
 
-			var/list/seen = viewers(4, get_turf(my_atom))
-			var/iconhtml = icon2html(cached_my_atom, seen)
-			if(cached_my_atom)
-				if(!ismob(cached_my_atom)) // No bubbling mobs
-					if(selected_reaction.mix_sound)
-						playsound(get_turf(cached_my_atom), selected_reaction.mix_sound, 80, TRUE)
-
-					for(var/mob/M in seen)
-						to_chat(M, "<span class='notice'>[iconhtml] [selected_reaction.mix_message]</span>")
-
-				if(istype(cached_my_atom, /obj/item/slime_extract))
-					var/obj/item/slime_extract/ME2 = my_atom
-					ME2.Uses--
-					if(ME2.Uses <= 0) // give the notification that the slime core is dead
-						for(var/mob/M in seen)
-							to_chat(M, "<span class='notice'>[iconhtml] \The [my_atom]'s power is consumed in the reaction.</span>")
-							ME2.name = "used slime extract"
-							ME2.desc = "This extract has been used up."
-
-			selected_reaction.on_reaction(src, multiplier)
-			reaction_occurred = TRUE
-			.++
+/*
+* Main Reaction loop handler, Do not call this directly
+*
+* Checks to see if there's a reaction, then processes over the reaction list, removing them if flagged
+* If any are ended, it displays the reaction message and removes it from the reaction list
+* If the list is empty at the end it finishes reacting.
+* Arguments:
+* * delta_time - the time between each time step
+*/
+/datum/reagents/process(delta_time)
+	if(!is_reacting)
+		finish_reacting()
+	//sum of output messages.
+	var/mix_message
+	//Process over our reaction list
+	//See equilibrium.dm for mechanics
+	for(var/datum/equilibrium/E in reaction_list)
+		SSblackbox.record_feedback("tally", "chemical_reaction", 1, "[E.reaction.type] total reaction steps")
+		//if it's been flagged to delete
+		if(E.to_delete)
+			mix_message += end_reaction(E)
+			continue
+		//otherwise continue reacting
+		E.react_timestep(delta_time)
+
+	if(mix_message)
+		my_atom.visible_message("<span class='notice'>[icon2html(my_atom, viewers(DEFAULT_MESSAGE_RANGE, src))] [mix_message]</span>")
+
+	if(!reaction_list.len)
+		finish_reacting()
+	else
+		update_total()
 
-	while(reaction_occurred)
+/*
+* This ends a single instance of an ongoing reaction
+*
+* Arguments:
+* * E - the equilibrium that will be ended
+* Returns:
+* * mix_message - the associated mix message of a reaction
+*/
+/datum/reagents/proc/end_reaction(datum/equilibrium/E)
+	//end reaction proc
+	E.reaction.reaction_finish(src, E.reacted_vol)
+	if(E.reaction.mix_sound)
+		playsound(get_turf(my_atom), E.reaction.mix_sound, 80, TRUE)
+	qdel(E)
+	reaction_list -= E
+	//Reaction occured
 	update_total()
-	if(.)
-		SEND_SIGNAL(src, COMSIG_REAGENTS_REACTED, .)
+	SEND_SIGNAL(src, COMSIG_REAGENTS_REACTED, .)
+	return E.reaction.mix_message
```

Comment:
In here you are still using the equilibrium datum even after it's been queued for deletion and ran its `Destroy()` proc.
This is not a good idea.
Clean-up the list before (or handle it in its `Destroy()` proc), and copy the `mix_message` into a local variable.

If this doesn't runtime `return E.reaction.mix_message` then that means that the equilibrium datum is keeping an unlawful reference to `reaction` that is not properly cleaned in `Destroy()`

Poke me in discord if you want a rundown on the garbage collector and handling references.